src/utils/safedir.py

```
@@ -0,0 +1,360 @@
+        """Release the held directory fd. Idempotent on double-exit.
+
+        Closing an already-closed fd would raise ``EBADF``, which is
+        noise in failure-path teardown; suppress it.
+        """
+        if self._closed:
+            return
+        self._closed = True
+        try:
+            os.close(self._fd)
+        except OSError:  # pragma: no cover - defensive
+            pass
```

Ruff (0.15.12)

[warning] Use `contextlib.suppress(OSError)` instead of `try`-`except`-`pass`

(SIM105)